presto-spi/src/main/java/com/facebook/presto/spi/memory/ArrayPool.java

```diff
@@ -0,0 +1,345 @@
+                long oldestAgeForAllObjects = Long.MIN_VALUE;
+                long totalAgeForAllObjects = 0;
+                int objectCountSnapshot = 0;
+                for (int i = 0; i < arrayObjectSizes.length; i++) {
+                    LongArrayList timeAddedForCurrentSize = timeAdded[i];
+                    ArrayList<T> objectList = arrayObjects[i];
+
+                    // synchronize on objectList to make sure no other threads would insert or remove from the objectList and timeAdded at the same time
+                    synchronized (objectList) {
+                        for (int j = 0; j < timeAddedForCurrentSize.size(); j += SAMPLING_FRACTION) {
+                            long ageForCurrentObject = now - timeAddedForCurrentSize.get(j);
+                            totalAgeForAllObjects += ageForCurrentObject;
+                            oldestAgeForAllObjects = max(oldestAgeForAllObjects, ageForCurrentObject);
+                            objectCountSnapshot++;
+                        }
+                    }
+                }
+                double averageAgeForAllObjects = (double) totalAgeForAllObjects / objectCountSnapshot;
+                long timeAddedThreshold = now - (long) ((oldestAgeForAllObjects + averageAgeForAllObjects) / 2);
+
+                // Evict objects who were added before or exactly at timeAddedThreshold, ie. timeAdded <= timeAddedThreshold
+                for (int i = 0; i < arrayObjectSizes.length; i++) {
+                    LongArrayList timeAddedForCurrentSize = timeAdded[i];
+                    ArrayList<T> objectList = arrayObjects[i];
+                    int currentSize = arrayObjectSizes[i];
+
+                    int sizeToRemove = 0;
+                    int cutoffIndex = 0;
+                    synchronized (objectList) {
+                        if (poolStats.get().totalSize <= capacity * 0.75) {
+                            trimming.set(false);
+                            return;
+                        }
+
+                        if (objectList.size() == 0) {
+                            continue;
+                        }
+
+                        int length = objectList.size();
+                        while (cutoffIndex < length && timeAddedForCurrentSize.get(cutoffIndex) <= timeAddedThreshold && poolStats.get().totalSize > capacity * 0.75) {
+                            objectList.remove(length - cutoffIndex - 1);
+                            cutoffIndex++;
+                        }
+
+                        if (cutoffIndex > 0) {
+                            timeAddedForCurrentSize.removeFront(cutoffIndex);
+                            sizeToRemove = cutoffIndex * currentSize;
+                        }
+                    }
+
+                    // Atomically update poolStats
+                    PoolStats currentPoolStats = poolStats.get();
```

Comment:
As Orri suggests, I think we can simplify this by relaxing the consistency constraint on `PoolStats`.  I think all of the fields on `PoolStats` can be atomic variables and we just increment/decrement on the fly.  This is for the use-case of exposing counters for ODS, where it's ok to report on the stats mid-update.  For the purpose of controlling the size of the `ArrayPool` via `totalSize`, we can have a separate atomic counter for that purpose which is more easily kept consistent.